meta/fraction-metadata/src/main/java/org/wildfly/swarm/fractions/ArquillianClientAnnotationSeekingClassVisitor.java

```diff
@@ -0,0 +1,100 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.swarm.fractions;
+
+import org.objectweb.asm.AnnotationVisitor;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * @author Juan Gonzalez
+ */
+public class ArquillianClientAnnotationSeekingClassVisitor extends ClassVisitor {
+
+    static final String DEPLOYMENT_ANNOTATION = "org/jboss/arquillian/container/test/api/Deployment";
+    static final String RUN_AS_CLIENT_ANNOTATION = "org/jboss/arquillian/container/test/api/RunAsClient";
+
+    private int clientCounter;
+    private int containerCounter;
+
+    public ArquillianClientAnnotationSeekingClassVisitor() {
+        super(Opcodes.ASM5);
+    }
+
+    public boolean isClient() {
+        return this.clientCounter > 0 && this.containerCounter == 0;
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int __,
+                                     final String ___,
+                                     final String desc,
+                                     final String signature,
+                                     final String[] exceptions) {
+
+        return new MethodVisitor(Opcodes.ASM5) {
+            @Override
+            public AnnotationVisitor visitAnnotationDefault() {
+                return new ArquillianAnnotationVisitor(desc);
+            }
+
+            @Override
+            public AnnotationVisitor visitAnnotation(final String desc,
+                                                     final boolean __) {
+
+                Type type = Type.getType(desc);
+                String className = type.getInternalName();
+
+                if (className.equals(DEPLOYMENT_ANNOTATION)) {
+                    containerCounter++;
+                    return new ArquillianAnnotationVisitor(className);
+                }
+
+                if (className.equals(RUN_AS_CLIENT_ANNOTATION)) {
+                    clientCounter++;
+                }
+
+                return null;
+            }
+       };
+    }
+
+    private final class ArquillianAnnotationVisitor extends AnnotationVisitor {
+            private String clazz;
+
+            public ArquillianAnnotationVisitor(String clazz) {
+                super(Opcodes.ASM5);
+                this.clazz = clazz;
+            }
+
+            @Override
+            public void visit(final String name, final Object value) {
+
+                if (clazz.equals(DEPLOYMENT_ANNOTATION) &&
+                        name.equals("testable")) {
+
+                    if (value.equals(Boolean.FALSE)) {
+                        clientCounter++;
+                        containerCounter--;
+                    } else {
+                        containerCounter++;
+                    }
```

Comment:
Personally, I don't quite understand the interactions between `ArquillianClientAnnotationSeekingClassVisitor` and `ArquillianAnnotationVisitor` around how the counters are incremented and decremented. It seems to be that `visitAnnotationDefault` doesn't apply to our situation, that method is called when looking at annotation definitions, not their usages, right? If so, I think this code could be simplified a bit.